syntax = "proto3";

message ExceptionResponse {
    string module_name = 1;
    string class_name = 2;
    string args = 3;
}

message ResourceRequest {
    string resource_type = 1;
    string client_uri = 2;
    repeated bytes args = 3;
}

message ResourceResponse {
    string handler = 1;
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    Status status = 2;
    ExceptionResponse exception = 3;
}

message ReleaseResourceRequest {
    string handler = 1;
    string client_uri = 2;
}

message ReleaseResourceResponse {
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    Status status = 2;
    ExceptionResponse exception = 3;
}

message RequestHeader {
    string version = 1;
    string handler = 2;
    string client_uri = 3;
    string method_name = 4;
}

message InstructionRequest {
    oneof instruction {
        RequestHeader header = 1;
        bytes message = 2;
    }
}

message ResponseHeader {
    float version = 1;
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    Status status = 2;
    ExceptionResponse exception = 3;
}

message InstructionResponse {
    oneof instruction {
        ResponseHeader header = 1;
        bytes message = 2;
    }
}

service RemoteTrackingService {
    rpc get_resource(ResourceRequest) returns (ResourceResponse) {}
    rpc release_resource(ReleaseResourceRequest) returns (ReleaseResourceResponse) {}
    rpc run_instruction(stream InstructionRequest) returns (stream InstructionResponse) {}
}
